File: CSharpCodeBase/components/character/starvation.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class starvation :component{
 var StarvationComponent = Class({__includes=Component, init=function(self, entity)
       Component.init(self, entity);
       this.entity.starvation = self;
       this.state = false;
       this.timeout = this.entity.config:Get("starvationTimeout")  ||  2;
       this.currentTimeout = this.timeout;
 }})
 public void StarvationComponent:SetState(state){
   this.state = state;
   this.currentTimeout = this.timeout;
 }
 public void StarvationComponent:Update(){
   Component.Update(self);
   if(this.state == false  &&  this.entity.health:GetState() != "low"  ){ 
     var multiplier = this.entity.health:GetState() == "high"  &&  1  ||  2;
     this.currentTimeout = this.currentTimeout - GameController.deltaTime * multiplier;
     if(this.currentTimeout < 0  ){ 
       this.currentTimeout = this.timeout;
       this.state = true;
     }
   }
 }
 public void StarvationComponent:OnRespawn(){
   self:Reset();
 }
 public void StarvationComponent:Reset(){
   self:SetState(false);
 }
 public void StarvationComponent:GetState(){
   return this.state  ||  this.entity.health:GetState() == "low";
 }
 Component}}

File: CSharpCodeBase/components/character/starvation.lua
local Class = require 'utils.hump.class'
local Component = require 'components.component'
local Config = require 'components.services.config'

local StarvationComponent = Class({__includes=Component, init=function(self, entity)
      Component.init(self, entity)
      self.entity.starvation = self
      self.state = false
      self.timeout = self.entity.config:Get('starvationTimeout') or 2
      self.currentTimeout = self.timeout
end})

function StarvationComponent:SetState(state)
  self.state = state
  self.currentTimeout = self.timeout
end

function StarvationComponent:Update()
  Component.Update(self)
  if self.state == false and self.entity.health:GetState() ~= 'low' then
    local multiplier = self.entity.health:GetState() == 'high' and 1 or 2
    self.currentTimeout = self.currentTimeout - GameController.deltaTime * multiplier
    if self.currentTimeout < 0 then
      self.currentTimeout = self.timeout
      self.state = true
    end
  end
end

function StarvationComponent:OnRespawn()
  self:Reset()
end

function StarvationComponent:Reset()
  self:SetState(false)
end

function StarvationComponent:GetState()
  return self.state or self.entity.health:GetState() == 'low'
end

return StarvationComponent